src/Compiler/Driver/ParseAndCheckInputs.fs

```diff
@@ -1427,126 +1433,407 @@ let CheckMultipleInputsSequential (ctok, checkForErrors, tcConfig, tcImports, tc
+[<RequireQualifiedAccess>]
+type NodeToTypeCheck =
+    /// A real physical file in the current project.
+    /// This can be either an implementation or a signature file.
+    | PhysicalFile of fileIndex: FileIndex
+    /// An artificial node that will add the earlier processed signature information to the TcEnvFromImpls.
+    /// Dependants on this type of node will perceive that a file is known in both TcEnvFromSignatures and TcEnvFromImpls.
+    /// Even though the actual implementation file was not type-checked.
+    | ArtificialImplFile of signatureFileIndex: FileIndex
+
+let folder (state: State) (finisher: Finisher<State, FinalFileResult>) : FinalFileResult * State = finisher.Invoke(state)
 
-        // Equip loggers to locally filter w.r.t. scope pragmas in each input
-        let inputsWithLoggers =
-            inputsWithLoggers
-            |> List.map (fun (input, oldLogger) ->
-                let logger = DiagnosticsLoggerForInput(tcConfig, input, oldLogger)
-                input, logger)
+/// Typecheck a single file (or interactive entry into F# Interactive)
+/// <returns>a callback functions that takes a `TcState` and will add the checked result to it.</returns>
+let CheckOneInputWithCallback
+    ((checkForErrors,
+      tcConfig: TcConfig,
+      tcImports: TcImports,
+      tcGlobals,
+      prefixPathOpt,
+      tcSink,
+      tcState: TcState,
+      inp: ParsedInput,
+      _skipImplIfSigExists: bool): (unit -> bool) * TcConfig * TcImports * TcGlobals * LongIdent option * TcResultsSink * TcState * ParsedInput * bool)
+    : Cancellable<Finisher<TcState, PartialResult>> =
+    cancellable {
+        try
+            CheckSimulateException tcConfig
 
-        // In the first linear part of parallel checking, we use a 'checkForErrors' that checks either for errors
-        // somewhere in the files processed prior to each one, or in the processing of this particular file.
-        let priorErrors = checkForErrors ()
+            let m = inp.Range
+            let amap = tcImports.GetImportMap()
 
-        // Do the first linear phase, checking all signatures and any implementation files that don't have a signature.
-        // Implementation files that do have a signature will result in a Choice2Of2 indicating to next do some of the
-        // checking in parallel.
-        let partialResults, (tcState, _) =
-            ((tcState, priorErrors), inputsWithLoggers)
-            ||> List.mapFold (fun (tcState, priorErrors) (input, logger) ->
-                use _ = UseDiagnosticsLogger logger
+            let conditionalDefines =
+                if tcConfig.noConditionalErasure then
+                    None
+                else
+                    Some tcConfig.conditionalDefines
+
+            match inp with
+            | ParsedInput.SigFile file ->
+                // printfn $"Processing Sig {file.FileName}"
```

Comment:
I don't have a strong view. If people often use Trace, then it might be worth cleaning up the printfn statements a bit and translating into Trace logs.